ml-agents/mlagents/trainers/components/reward_signals/reward_signal.py

```diff
@@ -0,0 +1,57 @@
+import logging
+from mlagents.trainers.trainer import UnityTrainerException
+from mlagents.trainers.policy import Policy
+
+import tensorflow as tf
+
+logger = logging.getLogger("mlagents.trainers")
+
+
+class RewardSignal(object):
+    def __init__(self, policy: Policy, strength, gamma):
+        """
+        Initializes a reward signal. At minimum, you must pass in the policy it is being applied to, 
+        the reward strength, and the gamma (discount factor.)
+        :param policy: The Policy object (e.g. PPOPolicy) that this Reward Signal will apply to. 
+        :param strength: The strength of the reward. The reward's raw value will be multiplied by this value. 
+        :param gamma: The time discounting factor used for this reward. 
+        :return: A RewardSignal object. 
+        """
+        raise UnityTrainerException(
+            "The initialize for this RewardSignal was not implemented."
+        )
+
+    def evaluate(self, current_info, next_info):
+        """
+        Evaluates the reward for the agents present in current_info given the next_info
+        :param current_info: The current BrainInfo. 
+        :param next_info: The BrainInfo from the next timestep.
+        :return: a tuple of (scaled intrinsic reward, unscaled intrinsic reward) provided by the generator
+        """
+        raise UnityTrainerException(
+            "The evaluate for this RewardSignal was not implemented."
+        )
+
+    def update(self, training_buffer, n_sequences):
+        """
+        If the reward signal has an internal model (e.g. GAIL or Curiosity), update that model.
+        :param training_buffer: The training buffer
+        :param n_sequences: The number of sequences in the training buffer.
+        :return: A dict of {"Stat Name": stat} to be added to Tensorboard
+        """
+        raise UnityTrainerException(
+            "The update for this RewardSignal was not implemented."
+        )
+
+    @classmethod
+    def check_config(cls, config_dict, param_keys=[]):
```

Comment:
It's probably OK here, but generally it's a bad idea to have mutable default arguments in python methods - see https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments.

A better way would be something like 
``` python
def check_config(cls, config_dict, param_keys=None):
    param_keys = param_keys or []
    for k in param_keys:
        # ...
```

Reply:
I think another way you could do this is something like
``` python
class RewardSignal(object):
  REQUIRED_PARAM_KEYS = []

  @classmethod
  def check_config(cls, config_dict):
  for k in cls.REQUIRED_CONFIG_PARAMS:
    # ...


class GAILRewardSignal(RewardSignal):
    REQUIRED_PARAM_KEYS = ["strength", "gamma", "demo_path"]
```
and then you wouldn't need to override check_config in each subclass.

Reply:
Got it - I'll fix this

Reply:
I went with the 1st method